crates/wasmparser/src/readers/core/operators.rs

```diff
@@ -423,16 +446,55 @@ pub trait VisitOperator<'a> {
                     $(
                         Operator::$op $({ $($arg),* })? => self.$visit($($($arg.clone()),*)?),
                     )*
+                    #[cfg(feature = "simd")]
+                    Operator::Simd(op) => {
+                        let Some(visitor) = self.simd_visitor() else {
+                            panic!("missing SIMD visitor for: {op:?}")
+                        };
+                        visitor.visit_simd_operator(op)
+                    }
                 }
             }
-
         }
         for_each_operator!(visit_operator)
     }
 
+    #[cfg(feature = "simd")]
+    fn simd_visitor(&mut self) -> Option<&mut dyn VisitSimdOperator<'a, Output = Self::Output>> {
+        None
+    }
```

Comment:
Could you add some documentation to this method as well? Mostly indicating what it is and how it's intended to be used. I think ideally it'd be good to have an example which shows the return-`Some(self)` trick

Reply:
Also, I'm not sure if it's possible, but one possible route of exploration would be to have an associated `type SimdVisitor;` for this trait where the return value is `Option<&mut Self::SimdVisitor>`. Typically users would write `type SimdVisitor = Self;`. The only slightly tricky part would be the default value of the associated type but I think that may be possible with something like `enum NonexistentVisitor {}` which can't be constructed.

Reply:
> Also, I'm not sure if it's possible, but one possible route of exploration would be to have an associated `type SimdVisitor;` for this trait where the return value is `Option<&mut Self::SimdVisitor>`. Typically users would write `type SimdVisitor = Self;`. The only slightly tricky part would be the default value of the associated type but I think that may be possible with something like `enum NonexistentVisitor {}` which can't be constructed.

This is how I would have probably designed this in the first place. However, your design idea has the crucial elegance of requiring no further boilerplate code for the default `NonExistentVisitor` which could be massive and diminish the compile time wins in the worst case.